hudi-common/src/main/java/org/apache/hudi/common/table/read/EngineBasedMerger.java

```diff
@@ -0,0 +1,194 @@
+      }
+    }
+  }
+
+  private static <T> BufferedRecord<T> getNewerRecordWithEventTimeOrdering(BufferedRecord<T> newer, BufferedRecord<T> older) {
+    if (newer.isCommitTimeOrderingDelete()) {
+      return newer;
+    }
+    if (older.isCommitTimeOrderingDelete()) {
+      return older;
+    }
+    Comparable newOrderingValue = newer.getOrderingValue();
+    Comparable oldOrderingValue = older.getOrderingValue();
+    return oldOrderingValue.compareTo(newOrderingValue) > 0 ? older : newer;
+  }
+
+  private Option<Pair<HoodieRecord, Schema>> getMergedRecord(BufferedRecord<T> older, BufferedRecord<T> newer) throws IOException {
+    ValidationUtils.checkArgument(!Objects.equals(payloadClass.get(), OverwriteWithLatestAvroPayload.class.getCanonicalName())
+        && !Objects.equals(payloadClass.get(), DefaultHoodieRecordPayload.class.getCanonicalName()));
+    HoodieRecord oldHoodieRecord = constructHoodieAvroRecord(readerContext, older);
+    HoodieRecord newHoodieRecord = constructHoodieAvroRecord(readerContext, newer);
+    return recordMerger.get().merge(oldHoodieRecord, getSchemaForAvroPayloadMerge(oldHoodieRecord, older),
+        newHoodieRecord, getSchemaForAvroPayloadMerge(newHoodieRecord, newer), props);
+  }
+
+  private Schema getSchemaForAvroPayloadMerge(HoodieRecord record, BufferedRecord<T> bufferedRecord) throws IOException {
+    if (record.isDelete(readerSchema, props)) {
+      return readerSchema;
+    }
+    return readerContext.getSchemaFromBufferRecord(bufferedRecord);
+  }
+
+  /**
+   * Constructs a new {@link HoodieAvroRecord} for payload based merging
+   *
+   * @param readerContext reader context
+   * @param bufferedRecord the provided engine specific record and its metadata
+   * @return A new instance of {@link HoodieRecord}.
+   */
+  private HoodieRecord constructHoodieAvroRecord(HoodieReaderContext<T> readerContext, BufferedRecord<T> bufferedRecord) {
+    GenericRecord record = null;
+    if (bufferedRecord.getRecord() != null) {
+      Schema recordSchema = readerContext.getSchemaFromBufferRecord(bufferedRecord);
+      record = readerContext.convertToAvroRecord(bufferedRecord.getRecord(), recordSchema);
+    }
+    return new HoodieAvroRecord<>(new HoodieKey(bufferedRecord.getRecordKey(), null), HoodieRecordUtils.loadPayload(payloadClass.get(), record, bufferedRecord.getOrderingValue()));
+  }
+
+  /**
+   * Picks the "latest" record based on the merger strategy and converts that to a delete by setting the `delete` flag.
+   * @param newer the newer record passed to the merge function
+   * @param older the older record passed to the merge function
+   * @return the latest record as a delete record
+   */
+  private BufferedRecord<T> getLatestAsDeleteRecord(BufferedRecord<T> newer, BufferedRecord<T> older) {
+    if (recordMerger.map(merger -> merger.getMergingStrategy().equals(HoodieRecordMerger.COMMIT_TIME_BASED_MERGE_STRATEGY_UUID)).orElse(false)) {
```

Comment:
So the record merge mode does not work perfectly here?

Reply:
This is to support the retraction message you had brought up. It avoids setting the value to null so you can send a retraction for the content. I never got a concrete example from you so I just went with my best understanding of the topic.

Reply:
It's tricky you emplify the msg payload first for delete before merger merging then try to recover it here, let's remove the special handling in `constructHoodieRecord` for deletes, just construct the record like others and merger would return the correct record based on the `orderingVal`, and the `BufferedRecord` can be recoverd correctly with right `HoodieOperation` and payload data(the row).

Reply:
The issue here is that the mergers will return an empty option when there is a delete: https://github.com/apache/hudi/blob/master/hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/HoodieSparkRecordMerger.java#L50

I think that this can cause some unexpected issues when using event time ordering since you will lose the ordering value that should be used when comparing to the next record.

For example, consider a delete at T2 and then it is followed by an insert with T1, that insert should be ignored but how will we keep track of that when the output drops this context?

Reply:
I've updated the code a bit and if either of the records are a delete we handle it first. This case where the merger outputs an empty option should not happen anymore but I am not sure of the safest way to handle this since a developer can provide any implementation of this merger logic.